src/c_kzg_4844.c

```diff
@@ -2369,344 +2370,150 @@ static C_KZG_RET ifft_fr(
+ * `CELLS_PER_EXT_BLOB` spaced root of unity in the domain.
  *
- * @param[out]  out             Polynomial with len_out fields allocated
- * @param[in]   len_out         Domain evaluation length, a power of two
- * @param[in]   scratch         Scratch space, at least 3x len_out
- * @param[in]   len_scratch     Scratch size, at least 3x len_out
- * @param[in]   partials        Array of polys to be multiplied together
- * @param[in]   partial_count   The number of polys to be multiplied together
- * @param[in]   s               The trusted setup
+ * @param[in,out]   vanishing_poly          The vanishing polynomial
+ * @param[in]       missing_cell_indices    The array of missing cell indices
+ * @param[in]       len_missing_cells       The number of missing cell indices
+ * @param[in]       s                       The trusted setup
  *
- * @remark This will pad the polynomials, perform FFTs, point-wise multiply the
- *         results together, and apply an inverse FFT to the result.
+ * @remark When all of the cells are missing, this algorithm has
+ * an edge case. We return C_KZG_BADARGS in that case.
+ * @remark When none of the cells are missing, recovery
+ * is trivial. We expect the caller to handle this case,
+ * and return C_KZG_BADARGS if not.
+ * @remark `missing_cell_indices` are assumed to be less than
+ * `CELLS_PER_EXT_BLOB`.
  */
-static C_KZG_RET reduce_partials(
-    poly_t *out,
-    size_t len_out,
-    fr_t *scratch,
-    size_t len_scratch,
-    const poly_t *partials,
-    size_t partial_count,
+static C_KZG_RET vanishing_polynomial_for_missing_cells(
+    fr_t *vanishing_poly,
+    const uint64_t *missing_cell_indices,
+    size_t len_missing_cells,
     const KZGSettings *s
 ) {
     C_KZG_RET ret;
+    fr_t *roots = NULL;
+    fr_t *short_vanishing_poly = NULL;
+    size_t short_vanishing_poly_len = 0;
 
-    /* Some sanity checks */
-    if (!is_power_of_two(len_out)) {
-        ret = C_KZG_BADARGS;
-        goto out;
-    }
-    if (len_scratch < len_out * 3) {
-        ret = C_KZG_BADARGS;
-        goto out;
-    }
-    if (partial_count == 0) {
+    /* Return early if none or all of the cells are missing */
+    if (len_missing_cells == 0 || len_missing_cells == CELLS_PER_EXT_BLOB) {
         ret = C_KZG_BADARGS;
         goto out;
     }
 
-    /*
-     * The degree of the output polynomial is the sum of the degrees of the
-     * input polynomials.
-     */
-    size_t out_degree = 0;
-    for (size_t i = 0; i < partial_count; i++) {
-        out_degree += partials[i].length - 1;
-    }
-    if (out_degree + 1 > len_out) {
+    /* Allocate arrays */
+    ret = new_fr_array(&roots, len_missing_cells);
+    if (ret != C_KZG_OK) goto out;
+    ret = new_fr_array(&short_vanishing_poly, (len_missing_cells + 1));
+    if (ret != C_KZG_OK) goto out;
+
+    /* Check if max_width is divisible by CELLS_PER_EXT_BLOB*/
+    if (s->max_width % CELLS_PER_EXT_BLOB != 0) {
         ret = C_KZG_BADARGS;
         goto out;
     }
```

Comment:
I feel that this particular check should be an assert, as it should never happen.